network/packethandlers.c

```diff
@@ -2,4 +2,598 @@
+    {
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+
+    errno = 0;
+    PTLVData_CONFIG_UPDATE_t* unpacked = calloc(1,sizeof(PTLVData_CONFIG_UPDATE_t));
+    //Allocation failed
+    if(unpacked == NULL && errno != 0)
+    {
+        //packet allocation succeeded, so free that memory
+        free(packet);
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+    packet->length = rawPacket->length;
+    packet->timestamp = rawPacket->timestamp;
+    packet->type = rawPacket->type;
+    packet->data = (PTLVData_Base_t*)unpacked;
+
+    //Note that this may be NULL if there's a failure
+    unpacked->new_pairs = parseKVList((char*)rawPacket->data, rawPacket->length);
+
+    llnet_packet_free(rawPacket);
+    return packet;
+}
+
+PacketTLV_t* unpackUserData(IntermediateTLV_t* rawPacket)
+{
+    PacketTLV_t* packet = createBasePacket(rawPacket);
+    if (packet == NULL)
+    {
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+
+    errno = 0;
+    PTLVData_USER_DATA_t* unpacked = calloc(1,sizeof(PTLVData_USER_DATA_t));
+    //Allocation failed
+    if(unpacked == NULL && errno != 0)
+    {
+        //packet allocation succeeded, so free that memory
+        free(packet);
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+
+    packet->data = (PTLVData_Base_t*)unpacked;
+
+    unpacked->left_stick_x = rawPacket->data[0];
+    unpacked->left_stick_y = rawPacket->data[1];
+    unpacked->right_stick_x = rawPacket->data[2];
+    unpacked->right_stick_y = rawPacket->data[3];
+    unpacked->button_a = (rawPacket->data[4] & 0x80) != 0;
+    unpacked->button_b = (rawPacket->data[4] & 0x40) != 0;
+    unpacked->controller_uuid = (((uint16_t)(rawPacket->data[6])) << 8) | rawPacket->data[7];
+
+    llnet_packet_free(rawPacket);
+    return packet;
+}
+
+PacketTLV_t* unpackDebug(IntermediateTLV_t* rawPacket)
+{
+    PacketTLV_t* packet = createBasePacket(rawPacket);
+    if (packet == NULL)
+    {
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+
+    errno = 0;
+    PTLVData_DEBUG_t* unpacked = calloc(1,sizeof(PTLVData_DEBUG_t));
+    //Allocation failed
+    if(unpacked == NULL && errno != 0)
+    {
+        //packet allocation succeeded, so free that memory
+        free(packet);
+        //Free the raw packet
+        llnet_packet_free(rawPacket);
+        return NULL;
+    }
+
+    packet->data = (PTLVData_Base_t*)unpacked;
+
+    unpacked->code_status = (rawPacket->data[0] & 0xF0) >> 4;
+    unpacked->commit_hash = (rawPacket->data[0] & 0x0F) << 24 | rawPacket->data[1] << 16| rawPacket->data[2] << 8 | rawPacket->data[3];
+    unpacked->robot_uuid = ((uint32_t*)(rawPacket->data))[1];
+    unpacked->state = (RobotState_t)(rawPacket->data[8]);
+    unpacked->config_entries = ((uint16_t*)(rawPacket->data))[5];
```

Comment:
I'm not a huge fan of magic number indexes, I'd rather see
```c
// note: not tested code
uint8_t* data_ptr= rawPacket->data;
unpacked->code_status = (data_ptr[0] & 0xf0) >> 4;
unpacked->commit_hash = ((data_ptr[0] & 0x0f) << 24) | (data_ptr[1] << 16) | (data_ptr[2] << 8) | (data_ptr[3] << 0);
data_ptr += 4;
unpacked->robot_uuid = ((uint32_t*) data_ptr)[0];
data_ptr += 4;
...
```

Reply:
Somewhat cleaner now